ISPN-12435 SingleKeyBackupWriteCommands run out of memory

danberindei opened the pull request:
https://issues.redhat.com/browse/ISPN-12435

* JGroupsTransport: complete request exceptionally instead of
  throwing an exception when marshalling fails.
* TriangleDistributionInterceptor: catch exception from JGroupsTransport
  when there is no request (sendToX methods).

pruivo commented:
@danberindei I don't think it solves the issue. 
If I understand the issue, the problem is in the backup:
* The primary owner tries to send the update `A` with `sequenceId` X but it fails (marshaling exception as an example)
* The next write, the primary owner sends successfully the update `B` with `sequenceId` X+1
* The backup ower won't handle the update `B` because it expects the `sequenceId` X to be delivered. 

We need some command to ignore sequence ids so the backup owner can make progress.

danberindei commented:
@pruivo I see, I haven't considered that the backups will not apply update X+1 until after they have applied update X. I'll have to look into why the tests pass, because I'm using the same caches for all 3 methods and I'm not doing anything with the `TriangleOrderManager` between methods. Maybe the test is just lucky and the keys are in different segments.

I suspect a similar problem appears when unmarshalling fails. Keys and values are unmarshalled before the command is passed to `TrianglePerCacheInboundInvocationHandler`, so an unmarshalling exception is going to be linked to the triangle sequence id on the originator, but not on the backup where unmarshalling failed. To fix it, we'd have to either delay the unmarshalling of keys and values until after `TrianglePerCacheInboundInvocationHandler`, or somehow let the originator know that the command was not executed at all and that it needs to send a "skip sequence id" command.

danberindei commented:
@pruivo I updated the test to check that a later put with the same key succeeds and I added a `BackupNoopCommand` that the primary sends when there was an error sending the command.

I split off the handling of unmarshalling errors on the remote node to a separate JIRA:
https://issues.redhat.com/browse/ISPN-12770

pruivo commented on core/src/test/java/org/infinispan/distribution/TriangleExceptionDuringMarshallingTest.java:
```diff
@@ -0,0 +1,220 @@
+      assertCleanFailure(originCache, backupKey);
+   }
+
+   @Test(enabled = false, description = "See ISPN-12770")
```
couldn't this be as simpler as catching and storing in `BackupWriteCommand` any exception from `readFrom()`?
then, when `invokeAsync()` is invoked, you would return a `CompletableFuture` with the exception: `return CompletableFutures.completedExceptionFuture(unmarshallException);`

danberindei replied:
Definitely not as simple as that, because the keys and values are read by the subclasses of `BackupWriteCommand` in their implementation of `readFrom()`. But I'll add a comment to ISPN-12770.

pruivo commented:
integrated! thanks @danberindei !
